### ISSUE TITLE
DuplicateNamesError when upgrading to 1.13

### ISSUE THREAD

**Issue by digitalmoksha** [rmosolgo/graphql-ruby#3919]
I'm in the process of upgrading to `1.13.8` (still coming from `1.11, and now skipping `1.12` since there is something in `1.13` we need).

Now right at the outset I'm getting `DuplicateNamesError`, which looks like comes from the new [Dynamic Fields](https://graphql-ruby.org/schema/dynamic_types.html) support that's been added:

```ruby
GraphQL::Schema::DuplicateNamesError:
  Found two visible type definitions for `CreateInput`: #<Class:0x00007fcafbb2a7f0>, #<Class:0x00007fcafb3cd908>
```

In fact, it's worse than that.  Setting a breakpoint in `graphql-1.13.8/lib/graphql/schema.rb`, I can see in `all_types['CreateInput']`:

```ruby
[#<Class:0x00007fcaedfaaba8>, #<Class:0x00007fca991f01d8>, #<Class:0x00007fca99145e90>, #<Class:0x00007fcac9d15dd0>, #<Class:0x00007fcabf52d438>, #<Class:0x00007fcabff52f30>, #<Class:0x00007fcabc5b93f0>, #<Class:0x00007fcabef41f88>]
```

This seems to stem from how we're creating our mutations.  Here's two examples relevant to the above:

```ruby
module Mutations
  module Issues
    class Create < BaseMutation
```

and

```ruby
module Mutations
  module Boards
    class Create < BaseMutation
```

This _seems_ to be the cause of generating multiple `CreateInput` types.

Is there any way around this without changing how we define our classes? Or am I overlooking what's going on?

**Comment by digitalmoksha** [comment]
Just a little more info. I can take a look at the arguments for the first two duplicates, and see that they are different.

```ruby
[1] pry(GitlabSchema)> all_types['CreateInput'][0].arguments
=> {"clientMutationId"=>#<GraphQL::Schema::Argument CreateInput.clientMutationId: String @description="A unique identifier for the client performing the mutation.">,
 "projectPath"=>#<Types::BaseArgument CreateBoard.projectPath: ID @description="Full path of the project with which the resource is associated.">,
 "groupPath"=>#<Types::BaseArgument CreateBoard.groupPath: ID @description="Full path of the group with which the resource is associated.">,
 "name"=>#<Types::BaseArgument CreateBoard.name: String @description="Board name.">,
 "hideBacklogList"=>#<Types::BaseArgument CreateBoard.hideBacklogList: Boolean @description="Whether or not backlog list is hidden.">,
 "hideClosedList"=>#<Types::BaseArgument CreateBoard.hideClosedList: Boolean @description="Whether or not closed list is hidden.">,
 "assigneeId"=>#<Types::BaseArgument CreateBoard.assigneeId: UserID @description="ID of user to be assigned to the board.">,
 "milestoneId"=>#<Types::BaseArgument CreateBoard.milestoneId: MilestoneID @description="ID of milestone to be assigned to the board.">,
 "iterationId"=>#<Types::BaseArgument CreateBoard.iterationId: IterationID @description="ID of iteration to be assigned to the board.">,
 "iterationCadenceId"=>#<Types::BaseArgument CreateBoard.iterationCadenceId: IterationsCadenceID @description="ID of iteration cadence to be assigned to the board.">,
 "weight"=>#<Types::BaseArgument CreateBoard.weight: Int @description="Weight value to be assigned to the board.">,
 "labels"=>#<Types::BaseArgument CreateBoard.labels: [String!] @description="Labels of the issue.">,
 "labelIds"=>#<Types::BaseArgument CreateBoard.labelIds: [LabelID!] @description="IDs of labels to be added to the board.">}
```

and

```ruby
[2] pry(GitlabSchema)> all_types['CreateInput'][1].arguments
=> {"clientMutationId"=>#<GraphQL::Schema::Argument CreateInput.clientMutationId: String @description="A unique identifier for the client performing the mutation.">,
 "description"=>#<Types::BaseArgument CreateIssue.description: String @description="Description of the issue.">,
 "dueDate"=>#<Types::BaseArgument CreateIssue.dueDate: ISO8601Date @description="Due date of the issue.">,
 "confidential"=>#<Types::BaseArgument CreateIssue.confidential: Boolean @description="Indicates the issue is confidential.">,
 "locked"=>#<Types::BaseArgument CreateIssue.locked: Boolean @description="Indicates discussion is locked on the issue.">,
 "type"=>#<Types::BaseArgument CreateIssue.type: IssueType @description="Type of the issue.">,
 "projectPath"=>#<Types::BaseArgument CreateIssue.projectPath: ID! @description="Project full path the issue is associated with.">,
 "iid"=>#<Types::BaseArgument CreateIssue.iid: Int @description="IID (internal ID) of a project issue. Only admins and project owners can modify.">,
 "title"=>#<Types::BaseArgument CreateIssue.title: String! @description="Title of the issue.">,
 "milestoneId"=>#<Types::BaseArgument CreateIssue.milestoneId: MilestoneID @description="ID of the milestone to assign to the issue. On update milestone will be removed if set to null.">,
 "labels"=>#<Types::BaseArgument CreateIssue.labels: [String!] @description="Labels of the issue.">,
 "labelIds"=>#<Types::BaseArgument CreateIssue.labelIds: [LabelID!] @description="IDs of labels to be added to the issue.">,
 "createdAt"=>#<Types::BaseArgument CreateIssue.createdAt: Time @description="Timestamp when the issue was created. Available only for admins and project owners.">,
 "mergeRequestToResolveDiscussionsOf"=>#<Types::BaseArgument CreateIssue.mergeRequestToResolveDiscussionsOf: MergeRequestID @description="IID of a merge request for which to resolve discussions.">,
 "discussionToResolve"=>#<Types::BaseArgument CreateIssue.discussionToResolve: String @description="ID of a discussion to resolve. Also pass `merge_request_to_resolve_discussions_of`.">,
 "assigneeIds"=>#<Types::BaseArgument CreateIssue.assigneeIds: [UserID!] @description="Array of user IDs to assign to the issue.">,
 "healthStatus"=>#<Types::BaseArgument CreateIssue.healthStatus: HealthStatus @description="Desired health status.">,
 "weight"=>#<Types::BaseArgument CreateIssue.weight: Int @description="Weight of the issue.">,
 "epicId"=>#<Types::BaseArgument CreateIssue.epicId: EpicID @description="ID of an epic to associate the issue with.">,
 "iterationId"=>#<Types::BaseArgument CreateIssue.iterationId: IterationID @description="Global iteration ID. Mutually exlusive argument with `iterationWildcardId`.">,
 "iterationWildcardId"=>
  #<Types::BaseArgument CreateIssue.iterationWildcardId: IssueCreationIterationWildcardId @description="Iteration wildcard ID. Supported values are: `CURRENT`. Mutually exclusive argument with `iterationId`. iterationCadenceId also required when this argument is provided.">,
 "iterationCadenceId"=>#<Types::BaseArgument CreateIssue.iterationCadenceId: IterationsCadenceID @description="Global iteration cadence ID. Required when `iterationWildcardId` is provided.">}
```

**Comment by rmosolgo** [comment]
Interesting -- somehow, the way to fix this will be to define `def self.default_graphql_name` in your `BaseMutation`, so that the generated input types use that correctly. But in order to do that, I need a little more context:

- When you use graphql 1.11, do you know what the names of these `*Input` are? (You could use `File.write("schema.graphql", MySchema.to_definition)` to get the whole schema printed out.) 
- Do you have any existing definitions of `def default_graphql_name`, `def self.default_graphql_name`, `def graphql_name` or `def self.graphql_name`? Maybe one of those has to be moved a bit (or maybe the gem broke one of them 😅 !)

In any case, something like this would _probably_ make it start working again, although it might not be backwards compatible, depending on what you find above: 

```ruby
module Mutations 
  class BaseMutation < ... # Whatever your BaseMutation does here 
    def self.default_graphql_name 
      _mutations_namespace, object_name, action_name = self.name.split("::")
      "#{action_name}#{object_name}"
    end 
  end 
end 
```

**Comment by rmosolgo** [comment]
Another possible difference is _when_ the GraphQL type name is configured. Previously, mutation input types weren't generated until mutations were actually attached to fields. But as part of 1.13, it was switched up to where those types are created as soon as the first `argument(...)` is configured. 

So, if there's any kind of custom naming code in your schema, it might have to be rejiggered to come _before_ any argument definitions.

**Comment by digitalmoksha** [comment]
We're not using implementing the `graphql_name` methods directly, but we set the `graphql_name` explicitly in the mutation. 

```ruby
module Mutations
  module Issues
    class Create < BaseMutation
      graphql_name 'CreateIssue'
```

and

```ruby
module Mutations
  module Boards
    class Create < BaseMutation
      graphql_name 'CreateBoard'
```

In `1.11`, this inputs fields are called `CreateBoardInput` and `CreateIssueInput`

**Comment by digitalmoksha** [comment]
I'll try the `self.default_graphql_name` override

**Comment by digitalmoksha** [comment]
Hmm, that does fix the duplication problem.  So it's now `CreateBoardsInput` and `CreateIssuesInput`

Of course these are plural versions of the `1.11` names, `CreateBoardInput` and `CreateIssueInput`

Is that itself going to cause backward incompatibility, the fact that the input names have changed?

**Comment by digitalmoksha** [comment]
> Another possible difference is when the GraphQL type name is configured. Previously, mutation input types weren't generated until mutations were actually attached to fields. But as part of 1.13, it was switched up to where those types are created as soon as the first argument(...) is configured.

We do include some common arguments in the definitions - moving the `graphql_name` above those seems to be giving us the correct, singular names!

**Comment by rmosolgo** [comment]
> moving the `graphql_name` above those

💰 Great, glad to hear it. 

For the record, yes, changing the name of a type is always a breaking change 😅 

If it turns out that just moving the name above the arguments isn't a good fix, I can probably investigate some other solution. (This same issue was reported here: https://github.com/rmosolgo/graphql-ruby/issues/3838#issuecomment-1019510752, and shuffling around some definitions did the job then, too.) 

Let me know if you run into any more trouble with it!

**Comment by digitalmoksha** [comment]
Thanks @rmosolgo for quick answer/solution!

**Comment by rmosolgo** [comment]
I _think_ that #3916  (to be released in 2.0.1) will address the problem described above, so that careful ordering of mutation configuration is not required anymore. If you happen to try it on that new version and find that the previous configuration still doesn't work, please let me know in a new issue! (Of course, if the fix from before is fine, no need to change things up on my account 🍻 )

**Comment by digitalmoksha** [comment]
I'm running into a few more instances of this.  Is https://github.com/rmosolgo/graphql-ruby/pull/3916 by chance on 1.13.11? I think you've only got it on 2.x.  Tried upgrading, it opens up other issues.
